src/compiler/factory/nodeFactory.ts

```diff
@@ -4481,6 +4482,24 @@ export function createNodeFactory(flags: NodeFactoryFlags, baseFactory: BaseNode
         return node;
     }
 
+    // @api
+    function updateModuleDeclarationAndKeyword(
+        node: ModuleDeclaration,
+        modifiers: readonly ModifierLike[] | undefined,
+        name: ModuleName,
+        body: ModuleBody | undefined
+    ) {
+        if (node.flags & (NodeFlags.Ambient | NodeFlags.Namespace)) {
+            return updateModuleDeclaration(node, modifiers, name, body);
+        }
+
+        return node.modifiers !== modifiers
+        || node.name !== name
+        || node.body !== body
+        ? update(createModuleDeclaration(modifiers, name, body, node.flags | NodeFlags.Namespace), node)
+        : update(createModuleDeclaration(node.modifiers, node.name, node.body, node.flags | NodeFlags.Namespace), node);

```

Comment:
Part of me wishes the `node.modifiers !== modifiers || node.name !== name || node.body !== body` wasn't a repeat of what we have in `updateModuleDeclaration` but also feels overkill to extract it out to a shared function and call out to it so I think overall having it the way you have it is good. The two functions are right next to each other so keeping them in-sync with any future changes should be clear to those future authors.

Nit: We can match the indentation similar to in `updateModuleDeclaration`

```suggestion
        return node.modifiers !== modifiers
            || node.name !== name
            || node.body !== body
            ? update(createModuleDeclaration(modifiers, name, body, node.flags | NodeFlags.Namespace), node)
            : update(createModuleDeclaration(node.modifiers, node.name, node.body, node.flags | NodeFlags.Namespace), node);

```

Reply:
Yes, I agree. Unfortunately I couldn't think of a better way to avoid duplicating this logic (aside from extracting it out as you said). We also agreed not to modify the (public) `updateModuleDeclaration` API as it seems safer to introduce an internal API instead that also updates the keyword where needed. Noted on the indentation, thanks!